### PR TITLE
 DataPipes and DataFrames integration

### PR DESCRIPTION
**Pull request by VitalyFedyunin**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #64309 [WIP] Add depth tracking
* **#60840 DataPipes and DataFrames integration**

Differential Revision: [D29461080](https://our.internmc.facebook.com/intern/diff/D29461080)

### CODE REVIEW

File: torch/utils/data/datapipes/iter/dataframes.py [hunk 1]
```diff
@@ -0,0 +1,416 @@
...
+        try:
+            return [i for i in item]
+        except:
+            return [item]
+
+    def __iter__(self):
+        aggregate = []
+        for item in self.source_datapipe:
+            aggregate.append(self._as_list(item))
+            if len(aggregate) == self.dataframe_size:
```
Comment (ejguan) [inline torch/utils/data/datapipes/iter/dataframes.py:145]: `len(aggregate)` can be larger than `dataframe_size` here. We need to yield first `dataframe_size` of items, and keep the rest in the buffer

File: torch/utils/data/datapipes/iter/dataframes.py [hunk 2]
```diff
@@ -0,0 +1,416 @@
...
+
+    def __iter__(self):
+        for item in self.source_datapipe:
+            yield self.output_var.calculate_me(item)
+
+
+@functional_datapipe('dataframes_as_tuples')
+class DataFramesAsTuplesPipe(IterDataPipe):
+    def __init__(self, source_datapipe):
+        self.source_datapipe = source_datapipe
```
Comment (ejguan) [inline torch/utils/data/datapipes/iter/dataframes.py:32]: Add assert that `isinstance(source_datapipe, DFIterDataPipe)`

File: torch/utils/data/datapipes/iter/dataframes.py [hunk 3]
```diff
@@ -0,0 +1,416 @@
...
+
+    def __iter__(self):
+        for df in self.source_datapipe:
+            for i in range(len(df.index)):
+                yield df[i:i + 1]
+
+
+@functional_datapipe('dataframes_concat', is_df=True)
+class ShuffleDataFramesPipe(DFIterDataPipe):
```
Comment (ejguan) [inline torch/utils/data/datapipes/iter/dataframes.py:52]: nit: ConcatDataFramesPipe

File: torch/utils/data/datapipes/iter/dataframes.py [hunk 4]
```diff
@@ -0,0 +1,416 @@
...
+            buffer.append(df)
+            if len(buffer) == size:
+                yield pandas.concat(buffer)
+                buffer = []
+        if len(buffer):
+            yield pandas.concat(buffer)
+
+
+@functional_datapipe('dataframes_filter', is_df=True)
+class ShuffleDataFramesPipe(DFIterDataPipe):
```
Comment (ejguan) [inline torch/utils/data/datapipes/iter/dataframes.py:97]: nit: FilterDataFramesPipe

File: torch/utils/data/datapipes/iter/dataframes.py [hunk 5]
```diff
@@ -0,0 +1,416 @@
+from torch.utils.data import IterDataPipe, functional_datapipe, DFIterDataPipe, DataChunk
+try:
+    import pandas
+    WITH_PANDAS = True
+finally:
```
Comment (ejguan) [inline torch/utils/data/datapipes/iter/dataframes.py:5]:
```suggestion
except ImportError:
```

**Review by ejguan — CHANGES_REQUESTED** [review]
A few comments for this PR.
And, do you think we should also align the APIs like `dataframe_shuffle` with `shuffle`? The stack of PRs has added like `unbatch` to the API, should we keep them same?

**Comment by VitalyFedyunin** [comment]
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29461080).

**Comment by VitalyFedyunin** [comment]
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29461080).

**Comment by VitalyFedyunin** [comment]
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29461080).

**Comment by VitalyFedyunin** [comment]
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29461080).

File: torch/utils/data/__init__.py [hunk 6]
```diff
@@ -2,8 +2,17 @@
...
-    (Dataset, IterableDataset, TensorDataset, ConcatDataset, ChainDataset,
-     Subset, random_split, Dataset as MapDataPipe, IterableDataset as IterDataPipe)
+    (ChainDataset,
+     ConcatDataset, 
+     DataChunk as DataChunk,
+     Dataset, 
+     Dataset as MapDataPipe,
+     DFIterDataPipe as DFIterDataPipe,
+     IterableDataset,
+     IterableDataset as IterDataPipe,
```
Comment (wenleix) [inline torch/utils/data/__init__.py:12]: nit: why `import IterableDataset` and `import IterableDataset as IterDataPipe`?

File: torch/utils/data/dataframes_pipes.ipynb [hunk 7]
```diff
@@ -0,0 +1,523 @@
...
+     "output_type": "stream",
+     "name": "stdout",
+     "text": [
+      "var_3 = input_var_2.i * 100\nvar_4 = var_3 + input_var_2.j\nvar_5 = var_4 - 2.7\ninput_var_2[\"y\"] = var_5\n"
+     ]
+    }
+   ],
+   "source": [
+    "dp = get_dataframes_pipe(dataframe_size = 3)\n",
+    "dp['y'] = dp.i * 100 + dp.j - 2.7\n",
```
Comment (wenleix) [inline torch/utils/data/dataframes_pipes.ipynb:146]: can i use `dp['i']` here as well ? e.g. 
```
dp['y'] = dp['i'] * 100 + dp['j'] - 2.7
```

File: torch/utils/data/datapipes/iter/dataframes.py [hunk 8]
```diff
@@ -0,0 +1,434 @@
...
+    def __iter__(self):
+        size = None
+        all_buffer = []
+        filter_res = []
+        for df in self.source_datapipe:
+            if size is None:
+                size = len(df.index)
+            for i in range(len(df.index)):
+                all_buffer.append(df[i:i + 1])
+                filter_res.append(self.filter_fn(df.iloc[i]))
```
Comment (wenleix) [inline torch/utils/data/datapipes/iter/dataframes.py:124]: it looks like `df.iloc[i]` returns the ith series (i.e. Column), while `df.iloc[[i]]` returns the ith row (confusing Pandas syntax, lol) . https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.iloc.html

File: torch/utils/data/dataframes_pipes.ipynb (hunk 7 above)
Comment (VitalyFedyunin) [reply torch/utils/data/dataframes_pipes.ipynb:146]: yes

File: torch/utils/data/__init__.py (hunk 6 above)
Comment (VitalyFedyunin) [reply torch/utils/data/__init__.py]: cleaned

File: torch/utils/data/datapipes/iter/dataframes.py (hunk 5 above)
Comment (VitalyFedyunin) [reply torch/utils/data/datapipes/iter/dataframes.py]: done

File: torch/utils/data/datapipes/iter/dataframes.py (hunk 4 above)
Comment (VitalyFedyunin) [reply torch/utils/data/datapipes/iter/dataframes.py]: done

File: torch/utils/data/datapipes/iter/dataframes.py (hunk 3 above)
Comment (VitalyFedyunin) [reply torch/utils/data/datapipes/iter/dataframes.py]: done

File: torch/utils/data/datapipes/dataframe/dataframes.py [hunk 9]
```diff
@@ -0,0 +1,294 @@
...
+        self.output_var = output_var
+
+    def __iter__(self):
+        for item in self.source_datapipe:
+            yield self.output_var.calculate_me(item)
+
+
+#  TODO(VitalyFedyunin): Extract this list from the DFIterDataPipe registred functions
+DATAPIPES_OPS = ['dataframes_as_tuples', 'groupby', 'dataframes_filter', 'map', 'to_datapipe',
+                 'shuffle', 'concat', 'batch', 'dataframes_per_row', 'dataframes_concat', 'dataframes_shuffle']
```
Comment (ejguan) [inline torch/utils/data/datapipes/dataframe/dataframes.py:24]: Are we exposing `dataframes_as_tuples`, `dataframes_per_row`, ... these API to users?

File: torch/utils/data/datapipes/dataframe/datapipes.py [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+    def __init__(self, source_datapipe):
+        self.source_datapipe = source_datapipe
+
+    def __iter__(self):
+        for df in self.source_datapipe:
+            for record in df.to_records(index=False):
+                yield record
+
+
+@functional_datapipe('dataframes_per_row', enable_df_api_tracing=True)
```
Comment (ejguan) [inline torch/utils/data/datapipes/dataframe/datapipes.py:28]: If these APIs are not user-facing, can we remove the functional API?

File: torch/utils/data/dataset.py [hunk 11]
```diff
@@ -82,12 +87,26 @@ def register_function(cls, function_name, function):
...
+            if isinstance(result_pipe, Dataset):
+                if enable_df_api_tracing or isinstance(source_dp, DFIterDataPipe) or getattr(result_pipe, '_dp_cast_to_df', False):
+                    if function_name not in UNTRACABLE_DATAFRAME_PIPES:
+                        result_pipe = result_pipe.trace_as_dataframe()
+
+                if getattr(result_pipe, '_dp_nesting_depth', None) is None:
+                    result_pipe._dp_nesting_depth = getattr(source_dp, '_dp_nesting_depth', None)
+
+                if getattr(result_pipe, '_dp_contains_dataframe', None) is None:
+                    result_pipe._dp_contains_dataframe = getattr(source_dp, '_dp_contains_dataframe', None)
```
Comment (ejguan) [inline torch/utils/data/dataset.py:106]: Could we add comments toward attribute of `_dp_cast_to_df`, `_dp_nesting_depth` and `_dp_contains_dataframe`?

File: test/test_datapipe.py [hunk 12]
```diff
@@ -343,6 +350,56 @@ def test_demux_mux_datapipe(self):
         self.assertEqual(list(range(10)), list(n))
 
 
+class TestDataFramesPipes(TestCase):
```
Comment (ejguan) [inline test/test_datapipe.py:353]: Is there way to test whether the Capture object is correct?

**Review by ejguan — COMMENTED** [review]
I think it looks reasonable to me with a few questions.
BTW, are we going to really land it into core? Can we move it into torchdata? Is there any dependency issue

**Comment by ejguan** [comment]
As long as we make all methods `Capture` to different DFIterDataPipe rather than embedded into normal IterDataPipe, I think we can untackle the dependency issue.

File: test/test_datapipe.py (hunk 12 above)
Comment (VitalyFedyunin) [reply test/test_datapipe.py:353]: For users they are printable (but with dynamic names), for code, execute and verify results probably the best approach.

**Comment by VitalyFedyunin** [comment]
> As long as we make all methods `Capture` to different DFIterDataPipe rather than embedded into normal IterDataPipe, I think we can untackle the dependency issue.

My think was to make DFPipes first class citizen to make sure all further development take them into account. As proposal we land them to core now, and work on decoupling after discovering all connection scenarios (there are few things remains to explore like tracking nesting depth properly, and behavior of standard pipes like grouping etc.)

File: torch/utils/data/datapipes/dataframe/dataframes.py (hunk 9 above)
Comment (VitalyFedyunin) [reply torch/utils/data/datapipes/dataframe/dataframes.py:24]: I will underscore them =)

File: torch/utils/data/datapipes/dataframe/dataframes.py [hunk 13]
```diff
@@ -0,0 +1,294 @@
...
+                guaranteed_group_size=None,
+                drop_remaining=False):
+        if unbatch_level != 0:
+            dp = self.unbatch(unbatch_level)._dataframes_per_row()
+        else:
+            dp = self._dataframes_per_row()
+        dp = dp.as_datapipe().groupby(group_key_fn, buffer_size=buffer_size, group_size=group_size,
+                                      guaranteed_group_size=guaranteed_group_size, drop_remaining=drop_remaining)
+        dp._dp_contains_dataframe = True
+        dp._dp_nesting_depth = 1
```
Comment (ejguan) [inline torch/utils/data/datapipes/dataframe/dataframes.py:111]: This is going to assign `_dp_nesting_depth` to 1

File: torch/utils/data/datapipes/iter/grouping.py [hunk 14]
```diff
@@ -65,7 +66,11 @@ def __init__(self,
         self.batch_size = batch_size
         self.drop_last = drop_last
         self.length = None
-        self.wrapper_class = DataChunk
+        source_depth = getattr(datapipe, '_dp_nesting_depth', 0)
+        if source_depth is None:
+            source_depth = 0
+        self._dp_nesting_depth = source_depth + 1
```
Comment (ejguan) [inline torch/utils/data/datapipes/iter/grouping.py:72]: We then increment `_dp_nesting_depth` from the normal `groupby`. Is this part really required?

**Review by ejguan — APPROVED** [review]
> My think was to make DFPipes first class citizen to make sure all further development take them into account. As proposal we land them to core now, and work on decoupling after discovering all connection scenarios 

That sounds great. We may need Wenlei to go over it again.

> (there are few things remains to explore like tracking nesting depth properly, and behavior of standard pipes like grouping etc.)

This is the part I am still not sure. Please see the inline comments.

File: torch/utils/data/datapipes/iter/grouping.py (hunk 14 above)
Comment (VitalyFedyunin) [reply torch/utils/data/datapipes/iter/grouping.py]: Moved all nesting to the next PR, as they still requires some cleanups.

File: torch/utils/data/datapipes/dataframe/dataframes.py (hunk 13 above)
Comment (VitalyFedyunin) [reply torch/utils/data/datapipes/dataframe/dataframes.py]: Moved all nesting to the next PR, as they still requires some cleanups.

File: torch/utils/data/datapipes/dataframe/dataframes.py [hunk 15]
```diff
@@ -0,0 +1,292 @@
...
+            self.ctx['variables'][0].source_datapipe, self)
+
+    def raw_iterator(self):
+        return self.as_datapipe().__iter__()
+
+    def __iter__(self):
+        return iter(self._dataframes_as_tuples())
+
+    def batch(self, batch_size=10):
+        dp = self._dataframes_per_row()._dataframes_concat(batch_size)
```
Comment (wenleix) [inline torch/utils/data/datapipes/dataframe/dataframes.py:91]: Curious on this: After `_dataframes_per_row` , the `DataPipe` will return one row a t a time (so it flatten the DataFrame`). 

But`_dataframes_concat`  will concat multiple DataFrames together?

**Review by wenleix — APPROVED** [review]
Thanks!

One high level question is whether there is any way to quickly identify a given `DataPipe` is a per-DF or per-row iterable? -- for example, sometimes it's not obvious whether the `source_datapipe` is a per-DF or per-row DataPipe without looking at the callsite, etc.

**Comment by VitalyFedyunin** [comment]
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29461080).

**Comment by VitalyFedyunin** [comment]
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29461080).

**Comment by VitalyFedyunin** [comment]
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29461080).

**Comment by VitalyFedyunin** [comment]
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29461080).
